storage/disk/txn.go

```diff
@@ -0,0 +1,423 @@
+// Copyright 2021 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package disk
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/dgraph-io/badger"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/internal/errors"
+	"github.com/open-policy-agent/opa/storage/internal/ptr"
+	"github.com/open-policy-agent/opa/util"
+)
+
+type transaction struct {
+	underlying *badger.Txn      // handle for the underlying badger transaction
+	partitions *partitionTrie   // index for partitioning structure in underlying store
+	pm         *pathMapper      // used for mapping between logical storage paths and actual storage keys
+	db         *Store           // handle for the database this transaction was created on
+	xid        uint64           // unique id for this transaction
+	stale      bool             // bit to indicate if the transaction was already aborted/committed
+	write      bool             // bit to indicate if the transaction may perform writes
+	context    *storage.Context // context supplied by the caller to be included in triggers
+}
+
+func newTransaction(xid uint64, write bool, underlying *badger.Txn, context *storage.Context, pm *pathMapper, trie *partitionTrie, db *Store) *transaction {
+	return &transaction{
+		underlying: underlying,
+		partitions: trie,
+		pm:         pm,
+		db:         db,
+		xid:        xid,
+		stale:      false,
+		write:      write,
+		context:    context,
+	}
+}
+
+func (txn *transaction) ID() uint64 {
+	return txn.xid
+}
+
+func (txn *transaction) Commit(_ context.Context) (storage.TriggerEvent, error) {
```

Comment:
[nit]
```suggestion
func (txn *transaction) Commit(context.Context) (storage.TriggerEvent, error) {
```

or rather a style question. It's not too important, I'll stop pointing them out repeatedly